### ISSUE TITLE
ATS 8.0.1 crash on ink_atomiclist_push

### ISSUE THREAD

**Issue by masaori335** [apache/trafficserver#4714]
```
[ 0  ] libc-2.12.so       __libc_waitpid                                         ( :undefined                ) 
[ 1  ] traffic_server     crash_logger_invoke                                    ( Crash.cc:165              ) 
[ 2  ] libc-2.12.so       0x2b87a7141570                                         ( :undefined                ) 
[ 3  ] libtscore.so.8.0.1 ink_atomiclist_push                                    ( ink_queue.cc:533          ) 
[ 4  ] traffic_server     ProtectedQueue::enqueue(Event*, bool)                  ( ProtectedQueue.cc:52      ) 
[ 5  ] traffic_server     EventProcessor::schedule(Event*, int, bool)            ( P_UnixEventProcessor.h:74 ) 
[ 6  ] traffic_server     transform_cleanup                                      ( HttpSM.cc:6707            ) 
[ 7  ] traffic_server     HttpSM::kill_this()                                    ( HttpSM.cc:6777            ) 
[ 8  ] traffic_server     HttpSM::main_handler(int, void*)                       ( HttpSM.cc:2557            ) 
[ 9  ] traffic_server     handleEvent                                            ( I_Continuation.h:160      ) 
[ 10 ] traffic_server     HttpTunnel::main_handler(int, void*)                   ( HttpTunnel.cc:1643        ) 
[ 11 ] traffic_server     handleEvent                                            ( I_Continuation.h:160      ) 
[ 12 ] traffic_server     write_signal_and_update                                ( UnixNetVConnection.cc:114 ) 
[ 13 ] traffic_server     write_signal_done                                      ( UnixNetVConnection.cc:156 ) 
[ 14 ] traffic_server     write_to_net_io                                        ( UnixNetVConnection.cc:507 ) 
[ 15 ] traffic_server     NetHandler::process_ready_list()                       ( UnixNet.cc:412            ) 
[ 16 ] traffic_server     NetHandler::waitForActivity(long)                      ( UnixNet.cc:528            ) 
[ 17 ] traffic_server     non-virtual thunk to NetHandler::waitForActivity(long) ( UnixNet.cc:undefined      ) 
[ 18 ] traffic_server     EThread::execute_regular()                             ( UnixEThread.cc:272        ) 
[ 19 ] traffic_server     spawn_thread_internal                                  ( Thread.cc:85              ) 
[ 20 ] libpthread-2.12.so start_thread                                           ( :undefined                )
```

**Comment by scw00** [comment]
Ho to re-product that ?

**Comment by masaori335** [comment]
I don't know how to reproduce. We see this on prod.
On report, not only ET_NET thread but also ACCEPT thread got signal.
```
[ 0  ] libc-2.12.so       __libc_waitpid                              ( :undefined                ) 
[ 1  ] traffic_server     crash_logger_invoke                         ( Crash.cc:165              ) 
[ 2  ] libc-2.12.so       0x2b1eaad22570                              ( :undefined                ) 
[ 3  ] libtscore.so.8.0.1 ink_atomiclist_push                         ( ink_queue.cc:533          ) 
[ 4  ] traffic_server     ProtectedQueue::enqueue(Event*, bool)       ( ProtectedQueue.cc:52      ) 
[ 5  ] traffic_server     EventProcessor::schedule(Event*, int, bool) ( P_UnixEventProcessor.h:74 ) 
[ 6  ] traffic_server     NetAccept::do_blocking_accept(EThread*)     ( UnixNetAccept.cc:346      ) 
[ 7  ] traffic_server     NetAccept::acceptLoopEvent(int, Event*)     ( UnixNetAccept.cc:524      ) 
[ 8  ] traffic_server     handleEvent                                 ( I_Continuation.h:160      ) 
[ 9  ] traffic_server     EThread::execute()                          ( UnixEThread.cc:317        ) 
[ 10 ] traffic_server     spawn_thread_internal                       ( Thread.cc:85              ) 
[ 11 ] libpthread-2.12.so start_thread                                ( :undefined                )
```

Something wrong with `ink_atomiclist_push`.

**Comment by scw00** [comment]
Which type of thread does it wants to schedule ?

**Comment by scw00** [comment]
The ProtectedQueue was initialized with EThread. So if Ethread initialized successfully, it shouldn't failed.

**Comment by vmamidi** [comment]
is this happening because of changes to event system? #2541

**Comment by masaori335** [comment]
@scw00 Probably, it's ET_TASK thread. Because `opt.etype` on frame 3 is `0`. But `e->ethread` on frame 2 is nullptr. It's assigned by `EventProcessor::assign_thread(EventType etype)` on P_UnixEventProcessor.h:68. Is it possible that it returns nullptr?

```
(gdb) bt
#0  ink_atomiclist_push (l=0x100b10, item=0x2acc8b9a3c40) at ink_queue.cc:533
#1  0x0000557b64c42233 in ProtectedQueue::enqueue (this=0x100b10, e=0x2acc8b9a3c40, fast_signal=true) at ProtectedQueue.cc:52
#2  0x0000557b64b73f76 in EventProcessor::schedule (this=<optimized out>, e=0x2acc8b9a3c40, etype=<optimized out>, fast_signal=true)
    at P_UnixEventProcessor.h:74
#3  0x0000557b64c1758e in NetAccept::do_blocking_accept (this=0x2acc37139280, t=0x2acc7dc0d0c0) at UnixNetAccept.cc:346
#4  0x0000557b64c167fb in NetAccept::acceptLoopEvent (this=0x2acc37139280, event=<optimized out>, e=<optimized out>) at UnixNetAccept.cc:524
#5  0x0000557b64c443b4 in Continuation::handleEvent (this=<optimized out>, event=1, data=0x1) at ./I_Continuation.h:160
#6  EThread::execute (this=0x2acc7dc0d0c0) at UnixEThread.cc:317
#7  0x0000557b64c42ecc in spawn_thread_internal (a=0x2acc37130500) at Thread.cc:85
#8  0x00002acc2dd43aa1 in start_thread () from /lib64/libpthread.so.0
#9  0x00002acc2da90c4d in clone () from /lib64/libc.so.6
(gdb) frame 3
#3  0x0000557b64c1758e in NetAccept::do_blocking_accept (this=0x2acc37139280, t=0x2acc7dc0d0c0) at UnixNetAccept.cc:346
346	    eventProcessor.schedule_imm_signal(vc, opt.etype);
(gdb) p opt.etype
$1 = 0
(gdb) frame 2
#2  0x0000557b64b73f76 in EventProcessor::schedule (this=<optimized out>, e=0x2acc8b9a3c40, etype=<optimized out>, fast_signal=true)
    at P_UnixEventProcessor.h:74
74	  e->ethread->EventQueueExternal.enqueue(e, fast_signal);
(gdb) p e->ethread
$2 = (EThread *) 0x0
```

**Comment by masaori335** [comment]
@vmamidi It looks like this is introduced by #1997.

**Comment by bryancall** [comment]
The fix has been cherry-picked into the 8.0.x branch fb0019bfed4e43a880872cfc3f7c63add9b04eed
